Feat/undo actions

abhifoyou opened the pull request:
This PR implements the Undo Action functionality and its related e2e test cases.

abhifoyou commented:
@jdpearce : Request you to review the same.

jdpearce commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -11,10 +12,83 @@ describe('When: I use the reading list feature', () => {
+
+    const enabledbooks = await $$('[data-testing="book-item"] button:not(:disabled)').first()
+
+    if (await enabledbooks.isPresent()) {
```
Similarly with this test - if any of these `if` statements are false, the test will "pass" but probably shouldn't.

abhifoyou replied:
i added the else part but please let me know if there is better way of doing this.

jdpearce replied:
See the other comment. I generally don't expect to see `if` statements in tests at all.

jdpearce commented on apps/okreads/browser/src/app/app.component.html:
```diff
@@ -23,7 +23,7 @@ <h1>okreads</h1>
       <h2>
         My Reading List
-        <button mat-icon-button (click)="drawer.close()" aria-label="Close the reading list Panel">
+        <button mat-icon-button (click)="drawer.close()" aria-label="Close the reading list Panel" id ="btnclose">
```
If you're adding an id for testing purposes only it's usually preferable to use a custom attribute instead, something like `data-testing="close-button"`.

abhifoyou replied:
Thanks for the suggestion. Changed the code.

jdpearce commented on libs/books/feature/src/lib/book-search/book-search.component.ts:
```diff
@@ -49,13 +53,37 @@ export class BookSearchComponent implements OnInit {
   }
 
+  undoAddToReadingList(item) {
+    //alert("inside undo add");
```
Did you mean to leave this in?

abhifoyou replied:
I am sorry. Somehow, forgot to remove the same.

Cleared. Thanks.

jdpearce commented on libs/books/feature/src/lib/book-search/book-search.component.ts:
```diff
@@ -49,13 +53,37 @@ export class BookSearchComponent implements OnInit {
 
   addBookToReadingList(book: Book) {
     this.store.dispatch(addToReadingList({ book }));
+    this.snackbar.openFromComponent(UndoActionSnackBarComponent, {
```
Should this really be here? What if the `addToReadingList` action fails?

abhifoyou replied:
I couldn't think of a better place to call the snack bar except this but i see what you are saying. Would you mind suggesting something on this ?

jdpearce replied:
The snackbar component should probably happen _after_ the confirm action. 

You could listen for some state change somewhere in this component and handle any kind of undo action. You would probably have to store the kind of action and how to undo it so that when you read it out you know what to dispatch when they click on the button...

abhifoyou replied:
I tried the same but somehow couldn't make it work. I will get back to you on this sometime later.

jdpearce commented on libs/books/feature/src/lib/reading-list/reading-list.component.html:
```diff
@@ -15,6 +15,7 @@
         color="warn"
         [attr.aria-label]="'Remove ' + b.title + ' from reading list'"
         (click)="removeFromReadingList(b)"
+        id= "removefromreading{{i}}"
```
Your formatting seems to be a bit inconsistent here and there. Some odd spaces seem to have crept in. Are you using Prettier?

abhifoyou replied:
Yes am using Preetier. Reformatted the same.

jdpearce commented on libs/books/feature/src/lib/reading-list/reading-list.component.ts:
```diff
@@ -10,9 +12,25 @@ import { getReadingList, removeFromReadingList } from '@tmo/books/data-access';
 
   removeFromReadingList(item) {
     this.store.dispatch(removeFromReadingList({ item }));
+    this.snackbar.openFromComponent(UndoActionSnackBarComponent, {
```
Similarly here, what if the `removeFromReadingList` action fails?

abhifoyou replied:
Same as the aforementioned related comment.

jdpearce commented on libs/books/feature/src/lib/reading-list/reading-list.component.ts:
```diff
@@ -10,9 +12,25 @@ import { getReadingList, removeFromReadingList } from '@tmo/books/data-access';
+    });
   }
+
+  undoToReadingList(item) {
```
Not sure about the name of this function...

abhifoyou replied:
undoFromReadingList would make more sense. Changed the same.

jdpearce commented on libs/books/feature/src/lib/undo-action-snackbar/undo-action-snackbar.component.html:
```diff
@@ -0,0 +1,9 @@
+    <span id="datamessage">{{data.message}}</span>
+    <div>
+        <button (click)="undoAction()" mat-button mat-icon-button id="undoaction"
+            aria-label="Undo Action">
```
Undo _what_ though?

abhifoyou replied:
Changed aria-label to be a generic one. We can do conditional checks too for exact/meaningful aria labels of required.

jdpearce commented on libs/books/feature/src/lib/undo-action-snackbar/undo-action-snackbar.component.ts:
```diff
@@ -0,0 +1,18 @@
+export class UndoActionSnackBarComponent {
+constructor(
+  public snackBarRef: MatSnackBarRef<UndoActionSnackBarComponent>,
+  @Inject(MAT_SNACK_BAR_DATA) public data: any) {}
```
Should this really be an `any`? Looks like it should at least be an `{ func: Function }`?

abhifoyou replied:
Sorry, i didn't get your point. Here i followed https://material.angular.io/components/snack-bar/overview which shows the similar code.

Would you please elaborate ?

jdpearce replied:
Typing the `data` property. It's not generally recommended to just leave it as an `any` type.

abhifoyou replied:
Got it. Changed the code accordingly . Thanks !

jdpearce reviewed (changes requested):
Quite a lot to clear up on this one. You may need to have a bit more of a think about how the undo behaviour works.

abhifoyou commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -11,10 +12,83 @@ describe('When: I use the reading list feature', () => {
-      )
+     )
     );
+    const closebutton = await browser.driver.findElement(by.id('btnclose')); //look for the undo button in the snackbar
```
That was a copy paste error. That button is for closing the reading list panel. 

Will clean the comment.

abhifoyou commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -11,10 +12,83 @@ describe('When: I use the reading list feature', () => {
+
+    const enabledbooks = await $$('[data-testing="book-item"] button:not(:disabled)').first() //looks for book that are not in reading list
+
+    if (await enabledbooks.isPresent()) {
```
I could write an else statement where i can declare an expectation. But my way of thinking was, if there are no books to add, then it doesn't make sense to go ahead with any expectations as we cannot test the undo functionality.

Please suggest a better way of doing this.

abhifoyou replied:
I hear you what are you saying, i added the else part but please let me know if there is better way of doing this.

jdpearce replied:
You've misunderstood me about the expectations here, I'd refactor this entire block to something like this:

```ts
    expect(enabledbooks.isPresent()).toBe(true);

    await enabledbooks.click();

    const totalCountElement = await element(by.tagName('tmo-total-count'));
    const undoaction = await browser.driver.findElement(by.id('undoaction'));
    await undoaction.click();

    expect(await totalCountElement.getText()).toBeTruthy();

    const count = await totalCountElement.element(
      by.css('span[class="mat-badge-content mat-badge-active"]')
    );
    let intCount = 0;
    const txt = await totalCountElement.getText();
    intCount = parseInt(txt, 10);

    const readListItemsCount = await element
      .all(by.css('.reading-list-item'))
      .count();
    expect(intCount).toEqual(readListItemsCount);
```

Note that you don't have to `await enabledbooks.isPresent()` when it's in the `expect` there.

abhifoyou commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -11,10 +12,83 @@ describe('When: I use the reading list feature', () => {
+    if (await enabledbooks.isPresent()) {
+      await enabledbooks.click();
+
+      const totalcountelement = await element(by.tagName('tmo-total-count'));
```
Sure, have changed the same. Thanks!

abhifoyou commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -11,10 +12,83 @@ describe('When: I use the reading list feature', () => {
+      const undoaction = await browser.driver.findElement(by.id('undoaction'));
+      await undoaction.click();
+
+      if (await totalcountelement.getText() !== "") {
```
As mentioned earlier, i added the else part but please let me know if there is better way of doing this.

abhifoyou commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -11,10 +12,83 @@ describe('When: I use the reading list feature', () => {
+        const txt = await totalcountelement.getText();
+        intCount = parseInt(txt, 10)
+
+        if (intCount > 0) {
```
You are right. This condition does not make sense. 

When i was coding the same, i was having multiple books in my reading list and hence it made me think of this logic probably. Have changed the code. Thanks.

abhifoyou commented:
@jdpearce : I have addressed most of the review comments but need few more inputs and suggestions from your side. Would you mind reviewing the comments please and suggest accordingly. Thanks!

jdpearce commented on libs/books/feature/src/lib/reading-list/reading-list.component.ts:
```diff
@@ -10,9 +12,25 @@ import { getReadingList, removeFromReadingList } from '@tmo/books/data-access';
+    });
   }
+
+  undoFromReadingList(item) {
```
Since this function isn't actually used by the template, you could make this private. Although it might make even more sense to just inline it, i.e.

```
   func: () => this.store.dispatch(addToReadingList({
          book: {
            id: item.bookId, 
            ...item
          }
        }))
```

abhifoyou replied:
Makes sense. Made it private.

jdpearce commented on libs/books/feature/src/lib/book-search/book-search.component.ts:
```diff
@@ -49,13 +53,36 @@ export class BookSearchComponent implements OnInit {
+});
   }
 
+  undoAddToReadingList(item) {
```
As with the comment below/above it might make more sense to inline this function. At the very least it's private.

abhifoyou replied:
Makes sense. Made it private.

jdpearce commented on libs/books/feature/src/lib/book-search/book-search.component.ts:
```diff
@@ -49,13 +53,36 @@ export class BookSearchComponent implements OnInit {
+        message: "Undo Add to reading List", 
+        action: "UNDO",
+        func: this.undoAddToReadingList.bind(this, book)},
+        duration: 3000
```
Is three seconds enough for a screen reader to find this?

abhifoyou replied:
Right, three second is actually allowing screen reader to find but its kind of overlapping everything and too fast. Made it 6 secs and looks ok to me. I tested using Chrome voice Screen reader tool.

jdpearce replied:
I'm not seeing that change here. Have you actually committed it?

abhifoyou replied:
Oops, i would have discarded the changes by mistake as i was working on the other review comments.

Re-added it.

jdpearce commented on libs/books/feature/src/lib/book-search/book-search.component.html:
```diff
@@ -8,7 +8,8 @@
     <button mat-icon-button matSuffix aria-label="Search Button for searching books to add to your reading list">
-      <mat-icon>search</mat-icon>
+      <mat-icon *ngIf="!searchTerm">search</mat-icon>
+      <mat-icon *ngIf="searchTerm" (click)="clearSearchTerm($event)" id="clearsearch">close</mat-icon>
```
Same point here about the `id`. Check the id's on all the tags, please. If you're only adding the id so it's accessible from the e2e tests then `id` is probably the wrong attribute to be using.

abhifoyou replied:
Sorry i missed this earlier. You already have suggested the same earlier in a different test case.

Fixed it now. Thanks.

jdpearce commented on libs/books/feature/src/lib/book-search/book-search.component.html:
```diff
@@ -8,7 +8,8 @@
     />
     <button mat-icon-button matSuffix aria-label="Search Button for searching books to add to your reading list">
-      <mat-icon>search</mat-icon>
+      <mat-icon *ngIf="!searchTerm">search</mat-icon>
```
Are you sure about this? Have you tried playing with this?

What you have here is now an icon that doesn't actually indicate the current state of the button. Try using this with a keyboard too - you'll notice that you cannot clear the search term (arguably you shouldn't be using the same button to do both actions anyway!)

abhifoyou replied:
I didn't think of this actually, thanks for pointing it out.

Have made it as different buttons.

jdpearce commented on libs/books/feature/src/lib/book-search/book-search.component.ts:
```diff
@@ -49,13 +53,36 @@ export class BookSearchComponent implements OnInit {
     this.searchBooks();
   }
 
+  clearSearchTerm(event) {
```
It looks like you're only really using this for the e2e tests. Why not just use `input.clear()`?

abhifoyou replied:
I am calling this in my template too so that user can clear the search term on click of close icon.

However, input.clear() would for work for e2e perfectly. Changed the code.

jdpearce commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -11,10 +12,91 @@ describe('When: I use the reading list feature', () => {
+      }
+    } else{
+      fail('Failed:I should be able to search and add the book in my reading list and then undo the same');
+      expect(true).toBe(false);
```
?

You've explicitly failed the test. Why do you need the expectation?

Rather than use if statements like this, a better solution would be to expect that the statement be true.

e.g. `expect(await totalCountElement.getText()).toBeTruthy();`

abhifoyou replied:
Absolutely Makes sense, This is what i was looking for. Thank you so much for educating me.

Changed the code.

abhifoyou commented:
@jdpearce : I have addressed all the review comments except the one in book-search.component.ts where "The snackbar component should probably happen after the confirm action." 
I tried making it work but i guess am missing something. I will get back to you on this after some time.

Please review again and let me know if i can move ahead with task 4 in parallel while i will be working to fix the aforementioned review suggestion too. Thanks !

jdpearce commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -1,8 +1,9 @@
 describe('When: I use the reading list feature', () => {
   it('Then: I should see my reading list', async () => {
     await browser.get('/');
+   //browser.ignoreSynchronization=true;
```
Did you mean to leave this here?

abhifoyou replied:
No, removed the same.

jdpearce commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -11,10 +12,84 @@ describe('When: I use the reading list feature', () => {
+     )
     );
+    const closebutton = await element(by.css('[data-testing="close-button"]'));
+    await closebutton.click();
```
Why are you closing the reading list here?

If you're doing this to reset the state, you could just put the `await browser.get('/');` in a `beforeEach`.

abhifoyou replied:
The first test case opens the Reading List panel and never closes it. So, while the next test case runs, it fails to find some of the underlying DOM elements so i had to add this line of code in the first test case of the test suite.

jdpearce commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -11,10 +12,84 @@ describe('When: I use the reading list feature', () => {
+
+      expect(intCount).toEqual(readListItemsCount);
+ 
+        const removefromreading = await element(by.id("removefromreading" + (readListItemsCount - 1)))
```
Your formatting seems to be getting out of hand here. Are you using Prettier?

abhifoyou replied:
Am using Prettier. Reformatted the same.

jdpearce commented on libs/books/feature/src/lib/book-search/book-search.component.html:
```diff
@@ -7,8 +7,11 @@
     />
-    <button mat-icon-button matSuffix aria-label="Search Button for searching books to add to your reading list">
-      <mat-icon>search</mat-icon>
+    <button *ngIf="!searchTerm" mat-icon-button matSuffix aria-label="Search Button for searching books to add to your reading list">
```
They're better than nothing, but these labels are _far_ too long. Something as simple as "Search Books" and "Clear Search" would be fine.

abhifoyou replied:
Changed the code.

jdpearce commented on libs/books/feature/src/lib/book-search/book-search.component.html:
```diff
@@ -41,6 +44,7 @@
               color="primary"
               (click)="addBookToReadingList(b)"
               [disabled]="b.isAdded"
+              id="wanttoreadbtn"
```
Should this be another `data-testing` attribute?

abhifoyou replied:
Am not using this anymore, added before but missed cleaning the same.

Removed the id.

jdpearce reviewed (changes requested):
A few more comments on this. I'll reserve final judgement until you've had a think about where to load the snackbar from. Let me know when you've figured out a better solution.

It also looks like you haven't committed all the changes that you've commented on, so I'll recheck these when your done on this task.

Feel free to tackle Task 4 concurrently while you're thinking about how to complete this one.

mandarini commented on libs/books/feature/src/lib/book-search/book-search.component.ts:
```diff
@@ -49,18 +54,52 @@ export class BookSearchComponent implements OnInit {
   }
 
+  clearSearchTerm(event) {
+    event.preventDefault();
```
Why are you preventing the default behavior here? What is the default behavior?

abhifoyou replied:
Removed the same.

mandarini commented on libs/books/feature/src/lib/book-search/book-search.component.ts:
```diff
@@ -1,45 +1,50 @@
 
   ngOnInit(): void {
-    this.store.select(getAllBooks).subscribe(books => {
+    this.store.select(getAllBooks).subscribe((books) => {
       this.books = books;
     });
```
I see now that this change should be in the code-review branch. And it seems that it is not. It seems that maybe you have not branched off of the code-review branch correctly, or you have not rebased. Can you update this please?

abhifoyou replied:
This was pointed out in Task 1 but intentionally not implemented as it was directed that we could implement/fix at least one of the issue mentioned. 

However, have changed the same now 👍

mandarini commented on libs/books/feature/src/lib/reading-list/reading-list.component.html:
```diff
@@ -1,7 +1,13 @@
+      <img
+        class="reading-list-item--cover"
+        [src]="b.coverUrl"
+        alt="{{ b.title }}"
```
Why are you using this syntax on this line, and the other syntax on the above line?

abhifoyou replied:
Fixed the same. Thanks for pointing it out.

mandarini commented on libs/books/feature/src/lib/reading-list/reading-list.component.html:
```diff
@@ -15,7 +21,7 @@
         [attr.aria-label]="'Remove ' + b.title + ' from reading list'"
         (click)="removeFromReadingList(b)"
-      >
+        id="removefromreading{{ i }}">
```
Same here, there is an inconsistency in the way you write your dynamic HTML attributes.

abhifoyou replied:
Fixed the same here too. Thanks for pointing it out.

mandarini reviewed (commented):
Your tests fail, unfortunately. Also, again, there are some syntax inconsistencies..

abhifoyou commented on libs/books/data-access/src/lib/+state/reading-list.reducer.ts:
```diff
@@ -2,13 +2,14 @@ import { Action, createReducer, on } from '@ngrx/store';
 
 import * as ReadingListActions from './reading-list.actions';
-import { ReadingListItem } from '@tmo/shared/models';
+import { ReadingListItem, SnackBarData } from '@tmo/shared/models';
```
Removed the same.

abhifoyou commented on libs/books/feature/src/lib/book-search/book-search.component.html:
```diff
@@ -7,8 +7,11 @@
-    <button mat-icon-button matSuffix aria-label="Search Button for searching books to add to your reading list">
-      <mat-icon>search</mat-icon>
+    <button *ngIf="!searchTerm" mat-icon-button matSuffix aria-label="Search Books">
+      <mat-icon >search</mat-icon>
```
Ran Prettier, should be formatted now.

abhifoyou commented on libs/books/feature/src/lib/book-search/book-search.component.ts:
```diff
@@ -1,45 +1,50 @@
-   }
+  }
 
   formatDate(date: void | string) {
```
Same as aforementioned comment. 

This was pointed out in Task 1 but intentionally not implemented as it was directed that we could implement/fix at least one of the issue mentioned.

However, have changed the same now 👍

abhifoyou commented:
@mandarini : I have addressed all the aforementioned review comments. We should be good now. Let me know in case of any discrepancy. 

Thanks for reviewing my PR :)

mandarini commented on libs/books/feature/src/lib/book-search/book-search.component.html:
```diff
@@ -7,35 +7,46 @@
       <div class="book--content">
         <div class="book--content--cover">
-          <img src="{{ b.coverUrl }}" alt="{{ b.title }}">
+          <img src="{{ b.coverUrl }}" alt="{{ b.title }}" />
```
Attribute binding syntax issue - see final comment

abhifoyou replied:
Changed the same, thanks.

mandarini commented on libs/books/feature/src/lib/reading-list/reading-list.component.html:
```diff
@@ -1,12 +1,13 @@
         class="reading-list-item--cover"
         [src]="b.coverUrl"
-        alt="{{ b.title }}"
+        [alt]="b.title"
```
Attribute binding syntax issue - see final comment

abhifoyou replied:
Changed the same, thanks.

mandarini commented on libs/books/feature/src/lib/reading-list/reading-list.component.html:
```diff
@@ -21,7 +22,8 @@
         [attr.aria-label]="'Remove ' + b.title + ' from reading list'"
         (click)="removeFromReadingList(b)"
-        id="removefromreading{{ i }}">
+        [attr.id]="'removefromreading' + i"
```
Attribute binding syntax here is correct - see final comment

mandarini commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -16,5 +16,83 @@ describe('When: I use the reading list feature', () => {
+    const undoaction = await browser.driver.findElement(by.id('undoaction'));
+    await undoaction.click();
+
+    expect(totalCountElement.getText()).toBeTruthy();
```
This test fails. The reason is that `totalCountElement.getText()` would be an empty string, most probably, if the state started off clean.

abhifoyou replied:
The test does not fail as am explicitly adding a new book to the reading list before and then executing the above step. Actually I was having an 'if' statement something like below and all the logic and the expect statements inside the if.
` if (await totalCountElement.getText() !== "") {`

But as suggested earlier by Johanna, we should avoid having if statements in e2e and rather we can use below to fail the test if element is not present (see comments above in this PR)
`expect(totalCountElement.getText()).toBeTruthy();`

And as am adding a book upfront, the test passes along with the test suite.

abhifoyou replied:
It should be good now as am explicitly adding a book to the reading list before doing anything and also changed the sequence here and there. It is passing now for the scenario when the state is clean. Thanks for pointing this out, i completely missed this scenario. 
![image](https://user-images.githubusercontent.com/25454641/95379881-f5be7500-08b3-11eb-9d3f-c37d23e3a6b6.png)

To give you a background, Actually I was having an 'if' statement something like below and all the logic and the expect statements inside the if.
`if (await totalCountElement.getText() !== "") {`
But as suggested earlier by Johanna, we should avoid having if statements in e2e and rather we can use below to fail the test if element is not present (see comments above in this PR)
`expect(totalCountElement.getText()).toBeTruthy();`

Anyways, we should be good now.

mandarini commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -16,5 +16,83 @@ describe('When: I use the reading list feature', () => {
+      by.css('span[class="mat-badge-content mat-badge-active"]')
+    );
+    let intCount = 0;
+    const txt = await totalCountElement.getText();
```
This test fails. The reason is that `totalCountElement.getText()`would be an empty string, most probably, if the state started off clean.

abhifoyou replied:
see above comment for the same file.

abhifoyou replied:
Should be good now. See my related comment above for the same file. Thanks for pointing out this scenario.

mandarini commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -16,5 +16,83 @@ describe('When: I use the reading list feature', () => {
+    );
+    let intCount = 0;
+    const txt = await totalCountElement.getText();
+    intCount = parseInt(txt, 10);
```
If `totalCountElement.getText()` is an empty string, then what would  `parseInt(txt, 10)` be?

abhifoyou replied:
see above comment for the same file.

abhifoyou replied:
Should be good now. See my related comment above for the same file. Thanks for pointing out this scenario.

mandarini commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -16,5 +16,83 @@ describe('When: I use the reading list feature', () => {
+    const readListItemsCount = await element
+      .all(by.css('.reading-list-item'))
+      .count();
+    expect(intCount).toEqual(readListItemsCount);
```
This fails, due to the reasons I explained above

abhifoyou replied:
see above comment for the same file.

abhifoyou replied:
Should be good now. See my related comment above for the same file. Thanks for pointing out this scenario.

mandarini reviewed (changes requested):
I think you have some trouble with the attribute binding syntax for Angular. I would suggest that you take a look at the documentation here: https://angular.io/guide/binding-syntax and here: https://angular.io/guide/attribute-binding and here: https://www.pluralsight.com/guides/attribute-class-style-bindings-angular!

I think this is an important part! You're very close, though! Good job!

You should also fix your e2e tests. Try to run the scenario manually, on your browser, and see what things you `expect` actually are.

abhifoyou commented:
@mandarini : I have addressed all the aforementioned review comments and e2e tests should be passing now for the scenario you mentioned. 

Thanks for reviewing my PR :)
